src/sysupdate/sysupdated.c

```diff
@@ -0,0 +1,1911 @@
+
+        SD_BUS_METHOD_WITH_ARGS("CheckNew",
+                                SD_BUS_NO_ARGS,
+                                SD_BUS_RESULT("s", new_version),
+                                target_method_check_new,
+                                SD_BUS_VTABLE_UNPRIVILEGED),
+
+        SD_BUS_METHOD_WITH_ARGS("Update",
+                                SD_BUS_ARGS("s", new_version),
+                                SD_BUS_RESULT("s", new_version, "u", job_id, "o", job_path),
+                                target_method_update,
+                                SD_BUS_VTABLE_UNPRIVILEGED),
+
+        SD_BUS_METHOD_WITH_ARGS("Vacuum",
+                                SD_BUS_NO_ARGS,
+                                SD_BUS_RESULT("u", count),
+                                target_method_vacuum,
+                                SD_BUS_VTABLE_UNPRIVILEGED),
+
+        SD_BUS_VTABLE_END
+};
+
+static const BusObjectImplementation target_object = {
+        "/org/freedesktop/sysupdate1/target",
+        "org.freedesktop.sysupdate1.Target",
+        .fallback_vtables = BUS_FALLBACK_VTABLES({target_vtable, target_object_find}),
+        .node_enumerator = target_node_enumerator,
+};
+
+static Manager *manager_free(Manager *m) {
+        if (!m)
+                return NULL;
+
+        hashmap_free(m->targets);
+        hashmap_free(m->jobs);
+
+        m->bus = sd_bus_flush_close_unref(m->bus);
+        sd_event_source_unref(m->notify_event);
+        sd_event_unref(m->event);
+
+        return mfree(m);
+}
+
+DEFINE_TRIVIAL_CLEANUP_FUNC(Manager *, manager_free);
+
+static int manager_on_notify(sd_event_source *s, int fd, uint32_t revents, void *userdata) {
+        char buf[NOTIFY_BUFFER_MAX+1];
+        struct iovec iovec = {
+                .iov_base = buf,
+                .iov_len = sizeof(buf)-1,
+        };
+        CMSG_BUFFER_TYPE(CMSG_SPACE(sizeof(struct ucred))) control;
+        struct msghdr msghdr = {
+                .msg_iov = &iovec,
+                .msg_iovlen = 1,
+                .msg_control = &control,
+                .msg_controllen = sizeof(control),
+        };
+        struct ucred *ucred;
+        Manager *m = ASSERT_PTR(userdata);
+        Job *j;
+        ssize_t n;
+        char *p;
+
+        n = recvmsg_safe(fd, &msghdr, MSG_DONTWAIT|MSG_CMSG_CLOEXEC);
+        if (n < 0) {
+                if (ERRNO_IS_TRANSIENT(n))
+                        return 0;
+                return (int) n;
+        }
+
+        cmsg_close_all(&msghdr);
+
+        if (msghdr.msg_flags & MSG_TRUNC) {
+                log_warning("Got overly long notification datagram, ignoring.");
+                return 0;
+        }
+
+        ucred = CMSG_FIND_DATA(&msghdr, SOL_SOCKET, SCM_CREDENTIALS, struct ucred);
+        if (!ucred || ucred->pid <= 0) {
+                log_warning("Got notification datagram lacking credential information, ignoring.");
+                return 0;
+        }
```

Comment:
one of these days we should probably introduce a common helper for reading notification messages like this. this is is kinda repetitive after all.